### ISSUE TITLE
Add polygonContain to Query

### ISSUE THREAD

**Issue by dplewis** [parse-community/parse-server#3942]
Mongo has a $geoIntersects that I want to use to return all polygons that contains a single point. Unfortunately it only supports 2dsphere indexes and parse-server currently doesn't use geoJson.

Maybe we can reopen https://github.com/parse-community/parse-server/pull/524 to support legacy coordinates (2d) as well as geoJSON (2dsphere) indexes.

Adding Type Polygon to Schema?

**Comment by flovilmart** [comment]
We need to introduce a new type to the schéma, i’m all Open for it as Long as we don’t break legacy points. We’ll also need to define those types in the client SDKs!

**Comment by dplewis** [comment]
Yeah I know its going to be a lot of work. Lets take it one step at a time.

**Comment by dplewis** [comment]
I've been using legacy points for years now. I'll be the first to know if it breaks.

For type polygon should it be formatted like.
```
{
    __type:'Polygon',
    coordinates: [[],[],[]]
}
```

**Comment by dplewis** [comment]
I've also read the 2dsphere indexes can cover multiple geospatial fields in a document. It might get rid of that MongoDB only supports one GeoPoint field in a class error

**Comment by flovilmart** [comment]
Yeah if using polygons obviously :)

**Comment by dplewis** [comment]
Maybe a tool to convert legacy to 2dsphere would help.

**Comment by flovilmart** [comment]
Yeah but that imply migrating data in production more likely, which could be problematic at large scale.

**Comment by dplewis** [comment]
The main issue I see is $nearSphere. I can test for 2d vs 2dsphere but the rest format are different. 

`2d`
```
location : { 
    $nearSphere : [ -73.9667, 40.78 ], 
    $maxDistance: 0.10 
}
```
`2dsphere`
```
location: {
    $nearSphere: {
        $geometry: {
              type : "Point",
              coordinates : [ -73.9667, 40.78 ]
        },
        $maxDistance: 0.10
    }
}
```

**Comment by flovilmart** [comment]
Also the indexes are different and I’m not sure you can index the same column 2 2D and 2dsphere

**Comment by dplewis** [comment]
You can have multiple geospatial indexes per collection. Unless you are using $geoNear which we aren't using.

https://docs.mongodb.com/manual/core/2dsphere/#geonear-and-geonear-restrictions

**Comment by dplewis** [comment]
I'll probably do 2dsphere indexes for legacy GeoPoints at a later date just closing this for now.

**Comment by madsb** [comment]
Could the $nearSphere dilemma be solved by adding the equivalent SDK methods for a $geoWithin query? We could have `withinCenterSphereRadians` and the equivalent `withinCenterSphereMiles` and `withinCenterSphereKilometers` methods. Then, for people who wants to use 2dsphere indexes, add in the documentation that calls that the legacy calls to `$nearSphere` are deprecated? I don't know about adding the indexes… maybe for starters we would have to do that manually?

**Comment by flovilmart** [comment]
@madsb do you wanna try adding a PR for it? Which version of mongodb do they start using those indexes?

**Comment by madsb** [comment]
I'll try to take a stab at it. I think $geoWithin queries are supported from version 2.4 and upwards: https://docs.mongodb.com/manual/reference/operator/query/geoWithin/

**Comment by dplewis** [comment]
@madsb `withinCenterSphere` doesn't sort like `$nearSphere` which a lot of applications have sorting. But not sorting leads to faster queries, we can use this as a option for users who don't care about sorting. To get `$nearSphere` to work with `2dsphere` is just changing the mongo transform and using `createIndexIfNeeded`

**Comment by madsb** [comment]
You're right about the sorting, I totally missed that (I'm not using sorting either in my app).

Still a bit novice about the whole index 2d/2dsphere thing. Databases are not my strong suit. Would probably be up to @drew-gross or @flovilmart to decide if now is the time to support GeoJSON. My vote is squarely in the 'yes' camp 😄

**Comment by dplewis** [comment]
I don't have time for PR because of work. But it's up for grabs. I can try to do a PR next week

**Comment by madsb** [comment]
I closed #4060. I may re-open it at a later point, but I made another pull request (#4064) that I'd prefer to have merged sooner, that allows for basic coordinate pairs in a $polygon query.
